server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -1108,4 +1038,160 @@ public static ClusterBlock createUUIDBasedBlock(ClusterBlock clusterBlock) {
+            AllocationService allocationService,
+            IndexMetadataVerifier indexMetadataVerifier,
+            IndicesService indicesService,
+            ShardLimitValidator shardLimitValidator
+        ) {
+            this.allocationService = allocationService;
+            this.indexMetadataVerifier = indexMetadataVerifier;
+            this.indicesService = indicesService;
+            this.shardLimitValidator = shardLimitValidator;
+        }
+
+        @Override
+        public ClusterTasksResult<OpenIndicesTask> execute(ClusterState currentState, List<OpenIndicesTask> tasks) throws Exception {
+            ClusterTasksResult.Builder<OpenIndicesTask> builder = ClusterTasksResult.builder();
+            ClusterState state = currentState;
+
+            try {
+                // build an in-order de-duplicated array of all the indices to open
+                final Set<Index> indicesToOpen = new LinkedHashSet<>(tasks.size());
+                for (OpenIndicesTask task : tasks) {
+                    for (Index index : task.request.indices()) {
+                        indicesToOpen.add(index);
+                    }
+                }
+                Index[] indices = indicesToOpen.toArray(new Index[0]);
+
+                // open them
+                state = openIndices(indices, state);
+
+                // do a final reroute
+                final String indicesAsString = Arrays.toString(indices);
+                state = allocationService.reroute(state, "indices opened [" + indicesAsString + "]");
+
+                for (OpenIndicesTask task : tasks) {
+                    builder.success(task);
+                }
+            } catch (Exception e) {
+                for (OpenIndicesTask task : tasks) {
+                    builder.failure(task, e);
+                }
+            }
+
+            return builder.build(state);
+        }
+
+        private ClusterState openIndices(final Index[] indices, final ClusterState currentState) {
+            final List<IndexMetadata> indicesToOpen = new ArrayList<>(indices.length);
+            for (Index index : indices) {
+                final IndexMetadata indexMetadata = currentState.metadata().getIndexSafe(index);
+                if (indexMetadata.getState() != IndexMetadata.State.OPEN) {
+                    indicesToOpen.add(indexMetadata);
+                } else if (currentState.blocks().hasIndexBlockWithId(index.getName(), INDEX_CLOSED_BLOCK_ID)) {
+                    indicesToOpen.add(indexMetadata);
+                }
+            }
+
+            shardLimitValidator.validateShardLimit(currentState, indices);
+            if (indicesToOpen.isEmpty()) {
+                return currentState;
+            }
+
+            logger.info(
+                () -> new ParameterizedMessage(
+                    "opening indices [{}]",
+                    String.join(",", indicesToOpen.stream().map(i -> (CharSequence) i.getIndex().toString())::iterator)
```

Comment:
This should have a limit. There's a risk that we run into a massive list here in edge cases and could create a multiple MB log message which is really annoying if it gets indexed or even just for scrolling to a file with `less` :)

Reply:
Yeah I was in two minds about this too. I decided not to say anything because (a) we already do this, (b) index names are not so long or numerous that we'd get scary-long log messages here and (c) I have definitely needed to grep for individual indices being opened/closed in support cases previously, it would have been a pain if they weren't mentioned in the logs due to truncation.

But since you also raised it I'm inclined to agree. Let's use `Strings#collectionToDelimitedStringWithLimit` here.

Reply:
ff2d068f056de6a4c623bb9b553f5b54a24d6dec